src/v2/SuperstateToken.sol

```diff
@@ -0,0 +1,463 @@
+
+    /// @notice Address of the AllowList contract which determines permissions for transfers
+    AllowList public immutable allowList;
+
+    /// @notice The next expected nonce for an address, for validating authorizations via signature
+    mapping(address => uint256) public nonces;
+
+    /// @notice Amount of an address's token balance that is encumbered
+    mapping(address => uint256) public encumberedBalanceOf;
+
+    /// @notice Amount encumbered from owner to taker (owner => taker => balance)
+    mapping(address => mapping(address => uint256)) public encumbrances;
+
+    /// @notice If all minting and burning operations are paused
+    bool public accountingPaused;
+
+    /// @notice Number of decimals used for the user representation of the token
+    uint8 private constant DECIMALS = 6;
+
+    /// @dev Event emitted when tokens are minted
+    event Mint(address indexed minter, address indexed to, uint256 amount);
+
+    /// @dev Event emitted when tokens are burned
+    event Burn(address indexed burner, address indexed from, uint256 amount);
+
+    /// @dev Emitted when the accounting pause is triggered by `admin`.
+    event AccountingPaused(address admin);
+
+    /// @dev Emitted when the accounting pause is lifted by `admin`.
+    event AccountingUnpaused(address admin);
+
+    /// @dev Thrown when a request is not sent by the authorized admin
+    error Unauthorized();
+
+    /// @dev Thrown when an address does not have sufficient permissions, as dictated by the AllowList
+    error InsufficientPermissions();
+
+    /// @dev Thrown when an address does not have a sufficient balance of unencumbered tokens
+    error InsufficientAvailableBalance();
+
+    /// @dev Thrown when the amount of tokens to spend or release exceeds the amount encumbered to the taker
+    error InsufficientEncumbrance();
+
+    /// @dev Thrown when the current timestamp has surpassed the expiration time for a signature
+    error SignatureExpired();
+
+    /// @dev Thrown if the signature has an S value that is in the upper half order.
+    error InvalidSignatureS();
+
+    /// @dev Thrown if the signature is invalid or its signer does not match the expected singer
+    error BadSignatory();
+
+    /// @dev Thrown if accounting pause is already on
+    error AccountingIsPaused();
+
+    /// @dev Thrown if accounting pause is already off
+    error AccountingIsNotPaused();
+
+    /// @dev Thrown if an address tries to encumber tokens to itself
+    error SelfEncumberNotAllowed();
+
+    /// @dev Thrown if array length arguments aren't equal
+    error InvalidArgumentLengths();
+
+    /**
+     * @notice Construct a new ERC20 token instance with the given admin and AllowList
+     * @param _allowList Address of the AllowList contract to use for permission checking
+     * @dev Disables initialization on the implementation contract
+     */
+    constructor(address _existingAdmin, AllowList _allowList) {
+        admin = _existingAdmin;
```

Comment:
Need to pass `_existingAdmin` in here to enforce that we are the only ones to call `initializeV2()`